### PR TITLE
(GH-860) Attempt all publishing tasks

### PR DESCRIPTION
**Pull request by gep13**
Relates to #860

### CODE REVIEW

**Comment by gep13** [comment]
@patriksvensson @devlead just putting this out there for discussion as a possible workaround to the problem that we show during the 0.11.0 release.  Thoughts?

File: build.cake [hunk 1]
```diff
@@ -207,6 +213,7 @@ Task("Publish-NuGet")
     .WithCriteria(() => parameters.IsMainCakeRepo)
     .WithCriteria(() => parameters.IsMainCakeBranch)
     .WithCriteria(() => parameters.IsTagged)
+    .ContinueOnError()
```
Comment (patriksvensson) [inline build.cake]: I think that `ContinueOnError` will suppress the `OnError` handling but I'm not sure.

File: build.cake [hunk 2]
```diff
@@ -317,7 +343,14 @@ Task("AppVeyor")
   .IsDependentOn("Publish-NuGet")
   .IsDependentOn("Publish-Chocolatey")
   .IsDependentOn("Publish-HomeBrew")
-  .IsDependentOn("Publish-GitHub-Release");
+  .IsDependentOn("Publish-GitHub-Release")
+  .Finally(() =>
```
Comment (patriksvensson) [inline build.cake]: This is a really smart use of `Finally` :smile: :+1:

File: build.cake (hunk 1 above)
Comment (gep13) [inline build.cake]: @patriksvensson yeah, I wasn't sure what would happen there.  Will need to test it out to see what happens.  If it does, any suggestions on how best to handle this?

Comment (patriksvensson) [inline build.cake]: @gep13 I think the easiest way to test this is a new Cake script with both `ContinueOnError` and `OnError` where you throw an exception in the task.

Comment (gep13) [inline build.cake]: @patriksvensson sounds like a plan, will give it a try, and report back.

**Comment by gep13** [comment]
@patriksvensson If I attempt the above in a stripped down version of the build script, I get the following:

image

Going to switch it out to use `ReportError` instead...

**Comment by gep13** [comment]
@patriksvensson @devlead ok, I "think" I am onto a winner :smile: 

What do you think about the following:

https://gist.github.com/gep13/028b4d3f3a98c3cbdfb122da9a35fdf8

**Comment by patriksvensson** [comment]
@gep13 ReportError won't work for this. Just use OnError and remove ContinueOnError. Read more here: http://cakebuild.net/docs/fundamentals/error-handling

**Comment by gep13** [comment]
>@patriksvensson said...
@gep13 ReportError won't work for this. Just use OnError and remove ContinueOnError. Read more here: http://cakebuild.net/docs/fundamentals/error-handling

Based on the test that I just did, it seems to work?!?

**Comment by gep13** [comment]
@patriksvensson if I remove ContinueOnError, how can I allow the build to continue onto the next task in the sequence?

**Comment by patriksvensson** [comment]
ReportError should do a`throw` and is meant to be a way to log errors but still keep the original call stack. If it doesn't work likw this then somwthing must be wrong :(

**Comment by gep13** [comment]
So my understanding of this is as follows...

I have a set of tasks which I want to run.  None of these tasks are dependent on each other, but I want to run them until, one after the other, and I want to control their execution order from a top level task.  To allow all the tasks to execute, regardless of whether each one succeeds or not, I use the `ContinueOnError` extension method for each task.  As a result, they all execute.

However, if one of these tasks does cause an error, I want to be able to Report, or rather, change a status flag that something has occurred.  I would have thought in this case that the ReportError extension method is doing exactly what it is supposed to.  Allowing an interception of the error that has occurred, without impacting on the overall flow.

What would you have expected to happen in the above scenario?

**Comment by patriksvensson** [comment]
@gep13 Yes. ReportError should have changed the status flag and then (internally) rethown the original exception. The OnError method (should be renamed to `Catch`) should do what you want.

**Comment by gep13** [comment]
Ah, ok, I see what you are saying...

So using this:

https://gist.github.com/gep13/cdf50d42aa7d4500bd43555013c72087

Also achieves the same thing.  Since we are not manually re-throwing the exception within OnError, it is as though the error never happened, and execution continues to the next task.

So, we might have an issue with ReportError, since it should have re-thrown.  However, I am wondering whether using ReportError in conjunction with ContinueOnError, actually results in the same functionality as OnError with no re-throw, i.e. silent continuation of workflow?

**Comment by patriksvensson** [comment]
@gep13 Ah, that's a side effect I haven't thought about! If this is the desired effect we should probably document it (both in documentation and with a unit and integration test).

**Comment by gep13** [comment]
>@patriksvensson said...
If this is the desired effect we should probably document it

Agreed.  If this shouldn't be allowed, then we should report an error, similar to when you use OnError and ContinueOnError on the same task.

**Comment by gep13** [comment]
@patriksvensson @devlead ok, I think this is ready for a review.

**Comment by patriksvensson** [comment]
Merged! :+1:

**Pull request merged**